src/codec/rtu.rs

```diff
@@ -238,60 +243,51 @@ fn decode<F>(
 where
     F: Fn(&BytesMut) -> Result<Option<usize>>,
 {
-    // TODO: Transform this loop into idiomatic code
-    loop {
-        let mut retry = false;
-        let res = get_pdu_len(buf)
-            .and_then(|pdu_len| {
-                debug_assert!(!retry);
+    const MAX_RETRIES: usize = 20;
+
+    for _i in 0..MAX_RETRIES {
+        match get_pdu_len(buf) {
+            Ok(pdu_len) => {
                 if let Some(pdu_len) = pdu_len {
-                    frame_decoder.decode(buf, pdu_len)
+                    return frame_decoder.decode(buf, pdu_len);
                 } else {
                     // Incomplete frame
-                    Ok(None)
+                    return Ok(None);
                 }
-            })
-            .or_else(|err| {
+            }
+            Err(err) => {
                 log::warn!("Failed to decode {} frame: {}", pdu_type, err);
                 frame_decoder.recover_on_error(buf);
-                retry = true;
-                Ok(None)
-            });
-        if !retry {
-            return res;
+            }
         }
     }
+
+    // Maximum number of retries exceeded.
+    log::error!("Giving up to decode frame after {} retries", MAX_RETRIES);
```

Comment:
`{MAX_RETRIES}`

Reply:
Changed. I refrained from doing that as I didn't know about your MSRV policy. The only hint was `edition = "2021"` indicating that I can use all features of Rust 1.56.
Capturing identifiers in format strings has increased the Minimum Supported Rust Version to 1.58.